File: Chapter-21/PassRun_1D.py
###########################################################################
### Chapter 21 - Football Decision Making 101                           ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

import pandas as pd
import numpy as np
import scipy.stats
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

dataset = pd.read_csv("pbp_1618.csv")
#epv = pd.read_csv("CWS.csv")

print("Pass-Run Ration on 1st and 10 at our own 25")
print(dataset["play_type"].value_counts())


#epa = [(epv[(epv.Down == 2) & (epv.Ydstogo == 10-dataset['yards_gained'].iloc[i])& (epv.Line == 25+dataset['yards_gained'].iloc[i])]["Value"].reset_index()['Value'][0]-epv[(epv.Down == 1) & (epv.Ydstogo == 10) & (epv.Line == 25)]["Value"].reset_index()['Value'][0]) if dataset['yards_gained'].iloc[i] < 10 else (epv[(epv.Down == 1) & (epv.Ydstogo == min(10,100-(25 + dataset['yards_gained'].iloc[i]))) & (epv.Line == 25 + dataset['yards_gained'].iloc[i])]["Value"].reset_index()['Value'][0]-epv[(epv.Down == 1) & (epv.Ydstogo == 10) & (epv.Line == 25)]["Value"].reset_index()['Value'][0]) for i in range(len(dataset))]


pass_epa = dataset[dataset.play_type == "pass"]["epa"]
run_epa = dataset[dataset.play_type == "run"]["epa"]

print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
print("The average EPA for a pass play is: ", np.mean(pass_epa), " with a success rate of: ", len([x for x in pass_epa if x>0])/len(pass_epa))
print("The average EPA for a run play is: ", np.mean(run_epa), " with a success rate of: ", len([x for x in run_epa if x>0])/len(run_epa))

matplotlib.rcParams.update({'font.size': 15})

plt.figure()
plt.subplot(211)
sns.distplot(pass_epa,label="Pass")
plt.yticks([0,0.2,0.4,0.6,0.8])
plt.xlabel("Expected Points Added")
plt.ylabel("Density")
plt.legend()

plt.subplot(212)
sns.distplot(run_epa,label="Run")
plt.yticks([0,0.2,0.4,0.6,0.8])
plt.xlabel("Expected Points Added")
plt.ylabel("Density")
plt.legend()

plt.savefig("Pass-Run-EPA.png")

File: Chapter-21/PassRun_1D.R
###########################################################################
### Chapter 21 - Football Decision Making 101                           ###
# Mathletics: How Gamblers, Managers, and Sports Enthusiasts              #
# Use Mathematics in Baseball, Basketball, and Football                   #
###########################################################################

library(readr)
library(ggplot2)
library(scales)

dataset <- read_csv("pbp_1618.csv")

cat("Pass-Run Ratio on 1st and 10 at our own 25:\n")
cat(table(dataset$play_type), "\n")

pass_epa <- dataset$epa[dataset$play_type == "pass"]
run_epa <- dataset$epa[dataset$play_type == "run"]

cat("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
cat("The average EPA for a pass play is:", mean(pass_epa), "with a success rate of:", sum(pass_epa > 0) / length(pass_epa), "\n")
cat("The average EPA for a run play is:", mean(run_epa), "with a success rate of:", sum(run_epa > 0) / length(run_epa), "\n")

# Plotting the distributions
ggplot(dataset, aes(x = epa, fill = play_type)) +
  geom_density(alpha = 0.5) +
  ylab("Density") +
  xlab("Expected Points Added") +
  scale_y_continuous(labels = percent) +
  facet_wrap(~ play_type, nrow = 2, scales = "free") +
  theme_minimal() +
  theme(legend.position = "none")

ggsave("Pass-Run-EPA.png", width = 8, height = 10, units = "in")

